File: sdks/javascript/src/credentials.ts
/**
 * Credential (W3C Verifiable Credentials) operations.
 */

import type { ATAPClient } from "./client.js";
import type { Credential } from "./models.js";

/** Email/phone/personhood verification and credential management. */
export class CredentialAPI {
  private readonly _client: ATAPClient;

  constructor(client: ATAPClient) {
    this._client = client;
  }

  /**
   * Initiate email verification (OTP). Requires atap:manage scope.
   *
   * @param email - Email address to verify.
   * @returns Status message.
   */
  async startEmailVerification(email: string): Promise<string> {
    const data = await this._client._authedRequest(
      "POST",
      "/v1/credentials/email/start",
      { jsonBody: { email } },
    );
    return (data.message as string) || "OTP sent";
  }

  /**
   * Verify email with OTP, issuing ATAPEmailVerification VC. Requires atap:manage scope.
   *
   * @param email - Email address.
   * @param otp - The OTP code.
   * @returns Credential with VC JWT.
   */
  async verifyEmail(email: string, otp: string): Promise<Credential> {
    const data = await this._client._authedRequest(
      "POST",
      "/v1/credentials/email/verify",
      { jsonBody: { email, otp } },
    );
    return parseCredential(data);
  }

  /**
   * Initiate phone verification (OTP). Requires atap:manage scope.
   *
   * @param phone - Phone number (E.164 format).
   * @returns Status message.
   */
  async startPhoneVerification(phone: string): Promise<string> {
    const data = await this._client._authedRequest(
      "POST",
      "/v1/credentials/phone/start",
      { jsonBody: { phone } },
    );
    return (data.message as string) || "OTP sent";
  }

  /**
   * Verify phone with OTP, issuing ATAPPhoneVerification VC. Requires atap:manage scope.
   *
   * @param phone - Phone number.
   * @param otp - The OTP code.
   * @returns Credential with VC JWT.
   */
  async verifyPhone(phone: string, otp: string): Promise<Credential> {
    const data = await this._client._authedRequest(
      "POST",
      "/v1/credentials/phone/verify",
      { jsonBody: { phone, otp } },
    );
    return parseCredential(data);
  }

  /**
   * Submit personhood attestation, issuing ATAPPersonhood VC. Requires atap:manage scope.
   *
   * @param options.providerToken - Optional provider token.
   * @returns Credential with VC JWT.
   */
  async submitPersonhood(
    options: { providerToken?: string } = {},
  ): Promise<Credential> {
    const body: Record<string, unknown> = {};
    if (options.providerToken) body.provider_token = options.providerToken;

    const data = await this._client._authedRequest(
      "POST",
      "/v1/credentials/personhood",
      { jsonBody: body },
    );
    return parseCredential(data);
  }

  /** List credentials for the authenticated entity. Requires atap:manage scope. */
  async list(): Promise<Credential[]> {
    const data = await this._client._authedRequest("GET", "/v1/credentials");
    if (Array.isArray(data)) {
      return (data as Record<string, unknown>[]).map(parseCredential);
    }
    return ((data.credentials as Record<string, unknown>[]) || []).map(
      parseCredential,
    );
  }

  /**
   * Get W3C Bitstring Status List VC (public endpoint).
   *
   * @param listId - Status list ID (default "1").
   */
  async statusList(
    listId = "1",
  ): Promise<Record<string, unknown>> {
    return this._client._http.request(
      "GET",
      `/v1/credentials/status/${listId}`,
    );
  }
}

export function parseCredential(data: Record<string, unknown>): Credential {
  return {
    id: (data.id as string) || "",
    type: (data.type as string) || "",
    credential: (data.credential as string) || "",
    issuedAt: (data.issued_at as string) || "",
    revokedAt: data.revoked_at as string | undefined,
  };
}

File: sdks/go/credentials.go
package atap

import "context"

// CredentialAPI provides email/phone/personhood verification and credential management.
type CredentialAPI struct {
	client *Client
}

// StartEmailVerification initiates email verification (OTP). Requires atap:manage scope.
func (c *CredentialAPI) StartEmailVerification(ctx context.Context, email string) (string, error) {
	data, err := c.client.authedRequest(ctx, "POST", "/v1/credentials/email/start", &RequestOptions{
		JSONBody: map[string]string{"email": email},
	})
	if err != nil {
		return "", err
	}
	msg := getString(data, "message")
	if msg == "" {
		msg = "OTP sent"
	}
	return msg, nil
}

// VerifyEmail verifies an email with OTP, issuing an ATAPEmailVerification VC. Requires atap:manage scope.
func (c *CredentialAPI) VerifyEmail(ctx context.Context, email, otp string) (*Credential, error) {
	data, err := c.client.authedRequest(ctx, "POST", "/v1/credentials/email/verify", &RequestOptions{
		JSONBody: map[string]string{"email": email, "otp": otp},
	})
	if err != nil {
		return nil, err
	}
	return parseCredential(data), nil
}

// StartPhoneVerification initiates phone verification (OTP). Requires atap:manage scope.
func (c *CredentialAPI) StartPhoneVerification(ctx context.Context, phone string) (string, error) {
	data, err := c.client.authedRequest(ctx, "POST", "/v1/credentials/phone/start", &RequestOptions{
		JSONBody: map[string]string{"phone": phone},
	})
	if err != nil {
		return "", err
	}
	msg := getString(data, "message")
	if msg == "" {
		msg = "OTP sent"
	}
	return msg, nil
}

// VerifyPhone verifies a phone with OTP, issuing an ATAPPhoneVerification VC. Requires atap:manage scope.
func (c *CredentialAPI) VerifyPhone(ctx context.Context, phone, otp string) (*Credential, error) {
	data, err := c.client.authedRequest(ctx, "POST", "/v1/credentials/phone/verify", &RequestOptions{
		JSONBody: map[string]string{"phone": phone, "otp": otp},
	})
	if err != nil {
		return nil, err
	}
	return parseCredential(data), nil
}

// SubmitPersonhood submits a personhood attestation, issuing an ATAPPersonhood VC. Requires atap:manage scope.
func (c *CredentialAPI) SubmitPersonhood(ctx context.Context, providerToken string) (*Credential, error) {
	body := map[string]interface{}{}
	if providerToken != "" {
		body["provider_token"] = providerToken
	}

	data, err := c.client.authedRequest(ctx, "POST", "/v1/credentials/personhood", &RequestOptions{JSONBody: body})
	if err != nil {
		return nil, err
	}
	return parseCredential(data), nil
}

// List lists credentials for the authenticated entity. Requires atap:manage scope.
func (c *CredentialAPI) List(ctx context.Context) ([]Credential, error) {
	data, err := c.client.authedRequest(ctx, "GET", "/v1/credentials", nil)
	if err != nil {
		return nil, err
	}

	var creds []Credential
	if items, ok := data["credentials"].([]interface{}); ok {
		for _, item := range items {
			if m, ok := item.(map[string]interface{}); ok {
				creds = append(creds, *parseCredential(m))
			}
		}
	}
	return creds, nil
}

// StatusList gets the W3C Bitstring Status List VC (public endpoint).
func (c *CredentialAPI) StatusList(ctx context.Context, listID string) (map[string]interface{}, error) {
	if listID == "" {
		listID = "1"
	}
	return c.client.http.Request(ctx, "GET", "/v1/credentials/status/"+listID, nil)
}

func parseCredential(data map[string]interface{}) *Credential {
	return &Credential{
		ID:         getString(data, "id"),
		Type:       getString(data, "type"),
		Credential: getString(data, "credential"),
		IssuedAt:   getString(data, "issued_at"),
		RevokedAt:  getString(data, "revoked_at"),
	}
}
